poc_entropy/src/multiplex_service.rs

```diff
@@ -0,0 +1,114 @@
+use axum::{body::BoxBody, http::header::CONTENT_TYPE, response::IntoResponse};
+use futures::{future::BoxFuture, ready};
+use hyper::{Body, Request, Response};
+use std::{
+    convert::Infallible,
+    task::{Context, Poll},
+};
+use tower::Service;
+
+pub struct MultiplexService<A, B> {
+    rest: A,
+    rest_ready: bool,
+    grpc: B,
+    grpc_ready: bool,
+}
+
+impl<A, B> MultiplexService<A, B> {
+    pub fn new(rest: A, grpc: B) -> Self {
+        Self {
+            rest,
+            rest_ready: false,
+            grpc,
+            grpc_ready: false,
+        }
+    }
+}
+
+impl<A, B> Clone for MultiplexService<A, B>
+where
+    A: Clone,
+    B: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            rest: self.rest.clone(),
+            grpc: self.grpc.clone(),
+            // the cloned services probably wont be ready
+            rest_ready: false,
+            grpc_ready: false,
+        }
+    }
+}
+
+impl<A, B> Service<Request<Body>> for MultiplexService<A, B>
+where
+    A: Service<Request<Body>, Error = Infallible>,
+    A::Response: IntoResponse,
+    A::Future: Send + 'static,
+    B: Service<Request<Body>, Error = Infallible>,
+    B::Response: IntoResponse,
+    B::Future: Send + 'static,
+{
+    type Response = Response<BoxBody>;
+    type Error = Infallible;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // drive readiness for each inner service and record which is ready
+        loop {
+            match (self.rest_ready, self.grpc_ready) {
+                (true, true) => {
+                    return Ok(()).into();
+                }
+                (false, _) => {
+                    ready!(self.rest.poll_ready(cx))?;
+                    self.rest_ready = true;
+                }
+                (_, false) => {
+                    ready!(self.grpc.poll_ready(cx))?;
+                    self.grpc_ready = true;
+                }
+            }
+        }
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        // require users to call `poll_ready` first, if they don't we're allowed to panic
+        // as per the `tower::Service` contract
+        assert!(
+            self.grpc_ready,
+            "grpc service not ready. Did you forget to call `poll_ready`?"
+        );
+        assert!(
+            self.rest_ready,
+            "rest service not ready. Did you forget to call `poll_ready`?"
+        );
+
+        // if we get a grpc request call the grpc service, otherwise call the rest service
+        // when calling a service it becomes not-ready so we have drive readiness again
+        if is_grpc_request(&req) {
+            self.grpc_ready = false;
+            let future = self.grpc.call(req);
```

Comment:
how does the grpc service handle this `call` method? is this something that's handled by tower?

Reply:
yes indeed.. this module was swiped from the Axum crate where they have it as an example

Reply:
which reminded me to add an attribution to the top of this file.

Reply:
Neat